cpp/include/raft/core/mdbuffer.cuh

```diff
@@ -0,0 +1,897 @@
+      case raft::memory_type::managed: result = create<raft::memory_type::managed>(res, n); break;
+      case raft::memory_type::pinned: result = create<raft::memory_type::pinned>(res, n); break;
+    }
+    return result;
+  }
+
+ private:
+  template <typename ContainerType>
+  auto static constexpr has_stream() -> decltype(std::declval<ContainerType>().stream(), bool())
+  {
+    return true;
+  };
+  auto static constexpr has_stream(...) -> bool { return false; };
+
+ public:
+  template <memory_type MemType>
+  [[nodiscard]] auto make_accessor_policy() noexcept
+  {
+    return accessor_policy<MemType>{};
+  }
+  template <memory_type MemType>
+  [[nodiscard]] auto make_accessor_policy() const noexcept
+  {
+    return const_accessor_policy<MemType>{};
+  }
+
+  [[nodiscard]] auto make_accessor_policy(memory_type mem_type) noexcept
+  {
+    auto result = accessor_policy_variant{};
+    switch (mem_type) {
+      case memory_type::host: result = make_accessor_policy<memory_type::host>(); break;
+      case memory_type::device: result = make_accessor_policy<memory_type::device>(); break;
+      case memory_type::managed: result = make_accessor_policy<memory_type::managed>(); break;
+      case memory_type::pinned: result = make_accessor_policy<memory_type::pinned>(); break;
+    }
+    return result;
+  }
+  [[nodiscard]] auto make_accessor_policy(memory_type mem_type) const noexcept
+  {
+    auto result = const_accessor_policy_variant{};
+    switch (mem_type) {
+      case memory_type::host: result = make_accessor_policy<memory_type::host>(); break;
+      case memory_type::device: result = make_accessor_policy<memory_type::device>(); break;
+      case memory_type::managed: result = make_accessor_policy<memory_type::managed>(); break;
+      case memory_type::pinned: result = make_accessor_policy<memory_type::pinned>(); break;
+    }
+    return result;
+  }
+};
+
+/**
+ * @brief A type representing multi-dimensional data which may or may not own
+ * its underlying storage. `raft::mdbuffer` is used to conveniently perform
+ * copies of data _only_ when necessary to ensure that the data are accessible
+ * in the desired memory space and format.
+ *
+ * When developing functions that interact with the GPU, it is often necessary
+ * to ensure that the data are in a particular memory space (e.g. device,
+ * host, managed, pinned), but those functions may be called with data that
+ * may or may not already be in the desired memory space. For instance, when
+ * called in one workflow, the data may have been previously transferred to
+ * device, rendering a copy unnecessary. In another, the function may be
+ * directly invoked on host data.
+ *
+ * Even when working strictly with host memory, it is often necessary to
+ * ensure that the data are in a particular layout for efficient access (e.g.
+ * column major vs row major) or that the the data are of a particular type
```

Comment:
There is an example of how to do this at line 246, but here is an example with a bit more commentary:
```
auto buf = raft::mdbuffer<float, raft::matrix_extent<int>, raft::row_major>{res, raft::mdbuffer{data}, raft::memory_type::device};
```
Here `data` is any mdspan with an `element_type` compatible with `float`. If `data` is column-major, an allocation will be made, and the data will be copied. In this case `buf` will be owning. If `data` is already a row-major `mdspan` of `float` on device, then `buf` will be non-owning, and no copy will be performed.

Reply:
Also, just noticed that the example at 246 was using `vector_extent` which is indeed misleading (since row/column major is the same for single-dimensional arrays). Pushing a one-word change to change it to `matrix_extent`.

Reply:
Final thought on this to explain the reasoning for the current design: One of the reasons it's good to avoid a named `transpose` utility function is that in the context of an `mdspan`, it is not clear what a `transpose` will actually do. Will it change the memory layout? Will it swap the indexing (in the sense of a matrix transpose)? Will it do both?

The way mdbuffer is currently written is to avoid questions like that because mdbuffer construction is always about "I want this data that already have, but I want to make sure it has _these_ mdspan properties." That may be memory location, data type, or layout, but what you get out is always some alternate representation of exactly what you put in.